### PR TITLE
Allow proxy mode server name to be configured

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently, proxy mode allows a remote cluster connection to be setup by
expecting all open connections to be routed through an intermediate
proxy. The proxy must use some logic to ensure that the connections end
up on the correct remote cluster. One mechanism provided is that the
default distribution TLS implementations will forward the host component
of the configured address to the remote connection using the SNI
extension. This is limiting as it requires that the proxy be configured
in a way that always uses a valid hostname as the proxy address.

Instead, this commit adds an additional setting to allow the server_name
to be configured independently. This allows the proxy address to be
specified as a IP literal, but the server_name specified as an arbitrary
string which still must be a valid hostname. It also decouples the
server_name from the requirement of being a DNS resolvable domain.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/ProxyConnectionStrategy.java [hunk 1]
```diff
@@ -75,12 +72,12 @@
...
     /**
-     * Whether to include the hostname as a server_name attribute
+     * A configurable server_name attribute
      */
-    public static final Setting.AffixSetting<Boolean> INCLUDE_SERVER_NAME = Setting.affixKeySetting(
+    public static final Setting.AffixSetting<String> SERVER_NAME = Setting.affixKeySetting(
         "cluster.remote.",
-        "include_server_name",
-        (ns, key) -> boolSetting(key, false, new StrategyValidator<>(ns, key, ConnectionStrategy.PROXY),
+        "server_name",
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/ProxyConnectionStrategy.java]: what's the upgrade path here?

**Review by ywelsch — COMMENTED** [review]
I've left one question

File: server/src/main/java/org/elasticsearch/transport/ProxyConnectionStrategy.java (hunk 1 above)
Comment (tbrooks8) [reply server/src/main/java/org/elasticsearch/transport/ProxyConnectionStrategy.java]: I'm not sure I understand the question here. Upgrade from a prior version? These settings are all new in 7.6.

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/transport/ProxyConnectionStrategy.java]: my bad, I did not see that this was in ProxyConnectionStrategy.

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**